### PR TITLE
[HOLD] Collapsing raw counts

### PR DESCRIPTION
**Pull request by yuhjong24**
Handle instances where index barcodes are shared.
- Added a parameter "sequencing_index_cols" to specify the sequencing related columns needed to uniquely identify every PCR wells.
- Changed the script and module for collate_fastq_reads to accommodate new parameter. 
- Changed the script and module for filtered_counts to accommodate new parameter.
- Changed read.csv to fread.
- Added some additional checks to collate_fastq_reads and filter_raw_reads.
- Added additional comments

### CODE REVIEW

File: scripts/src/collate_fastq_reads.R [hunk 1]
```diff
@@ -4,39 +4,94 @@
...
+  # "flowcell_lane" is a string containing a group of lanes, to be parsed by splitting on the chars , ; :
+  # "single_lane" contains the individual lane as a number.
   # Note: fread and read.csv keeps commas, read_csv DROPS commas
   meta_flowcells= sample_meta %>% dplyr::distinct(flowcell_name, flowcell_lane) %>%
-    dplyr::mutate(flowcell_lane= base::strsplit(flowcell_lane, split='[,;:]', fixed=F)) %>%
-    tidyr::unnest(cols= flowcell_lane) %>%
-    dplyr::mutate(flowcell_lane= as.numeric(flowcell_lane))
+    dplyr::mutate(single_lane= base::strsplit(flowcell_lane, split='[,;:]', fixed=F)) %>% 
+    tidyr::unnest(cols= single_lane) %>%
+    dplyr::mutate(single_lane= as.numeric(single_lane))
```
Comment (AnupJonchhe) [inline scripts/src/collate_fastq_reads.R:33]: Do the same with flowcell_names

**Review by AnupJonchhe — CHANGES_REQUESTED** [review]

File: scripts/src/filter_raw_reads.R [hunk 2]
```diff
@@ -24,6 +24,9 @@ suppressPackageStartupMessages(library(sets))
 #'                - Sequence
 #'                - Name
 #'                - log_dose
+#' @param sequencing_index_cols Vector of column names from the sample meta that is used to uniquely identify where a
```
[the remark this answers is not in the document]
Comment (yuhjong24) [reply scripts/src/filter_raw_reads.R:27]: Yes, so normally a sample (one PCR well) is sequenced across several lanes. Usually its across 4 lanes. The output of nori has the reads split by the individual lanes. This module sums the reads across the four lanes by grouping on the index barcodes and the read sequence. But with recent runs, there are different samples with the same index barcodes, but they were sequenced with different flowcells. So in these cases we do not want to collapse across flowcell_names and/or flowcell_lanes.

File: scripts/src/collate_fastq_reads.R [hunk 3]
```diff
@@ -4,39 +4,94 @@
 #' filters for reads coming from flowcells specificed in the sample meta.
 #' The function then sums up the reads the flowcells using the index pairs.
 #' 
-#' @param sample_meta Sample metadata generate for the project
 #' @param uncollapsed_raw_counts Data frame of reads from all the fastq files with the following columns - \cr
-#'                    "flowcell_name", "flowcell_lane", "index_1", "index_2", and "n"
+#'                    "flowcell_name", "flowcell_lane", "index_1", "index_2", and "forward_read_cl_barcode", "n"
+#' @param sample_meta Sample metadata generate for the project  
+#' @param sequencing_index_cols Sequencing columns from the sample meta that the counts should be collapsed on. \cr
```
[the remark this answers is not in the document]
Comment (yuhjong24) [reply scripts/src/collate_fastq_reads.R:10]: In a sense yes - if "IndexBarcode1,IndexBarcode2" are provided, the module will group by those AND the "forward_read_cl_barcode" and sum up the number of times a read appears across the flowcells. The collapse will always include "forward_read_cl_barcode". The resulting file will have the columns - "index_1", "index_2", "forward...", and "n".

**Comment by ashishge1** [comment]
Just wanted to note that in the new code comments, to use "validation" instead of "QC" to not confuse with other pipeline QC.

**Review by maha-naim — APPROVED** [review]

File: R/QC_images.R [hunk 4]
```diff
@@ -101,14 +128,14 @@ QC_images = function(sample_meta, raw_counts, annotated_counts, normalized_count
...
-    merge(cell_set_meta, by="cell_set", all.x=T) %>%
-    dplyr::mutate(members= ifelse(is.na(members), cell_set, members), # for custom cell sets
-                  expected_num_cl = as.character(members) %>% purrr::map(strsplit, ";") %>% 
+    dplyr::left_join(cell_set_meta, by="cell_set", relationship= 'many-to-one') %>%
+    dplyr::mutate(members= if_else(is.na(members), cell_set, members), # for custom cell sets
+                  expected_num_cl= as.character(members) %>% purrr::map(strsplit, ";") %>% 
                     purrr::map(`[[`, 1) %>% purrr::map(length) %>% as.numeric(),
                   count_type= ifelse(n > count_threshold, 'Detected', 'Low'),
                   count_type= ifelse(n==0, 'Missing', count_type)) %>%
     group_by(profile_id, pcr_plate, pcr_well, count_type, expected_num_cl) %>%
```
Comment (manilkocak) [inline R/QC_images.R:137]: These two lines can be replaced with dplyr::count(profile_id, pcr_plate, pcr_well, count_type, expected_num_cl)

File: R/QC_images.R [hunk 5]
```diff
@@ -299,7 +410,7 @@ QC_images = function(sample_meta, raw_counts, annotated_counts, normalized_count
         filter(!is.na(CCLE_name),
                (!trt_type %in% c("empty", "", "CB_only")) & !is.na(trt_type)) %>% 
         mutate(plt_id= paste(sig_id, bio_rep, sep=':')) %>% 
-        dcast(CCLE_name~plt_id, value.var="mean_normalized_n") %>% 
+        reshape2::dcast(CCLE_name~plt_id, value.var="mean_normalized_n") %>%
```
Comment (manilkocak) [inline R/QC_images.R:413]: if you replace dcast with acast you won't need the following line.

File: scripts/src/filter_raw_reads.R [hunk 6]
```diff
@@ -38,47 +73,77 @@ suppressPackageStartupMessages(library(sets))
...
   CB_meta= CB_meta %>% dplyr::mutate(log2_dose= log_dose/log10(2)) %>% dplyr::select(-log_dose)
   
   if (reverse_index2) {
-    print("Reverse-complementing index 2 barcode ...")
-    sample_meta$IndexBarcode2 <- chartr("ATGC", "TACG", stringi::stri_reverse(sample_meta$IndexBarcode2))
+    if ('index_2' %in% colnames(sample_meta)) {
+      print("Reverse-complementing index 2 barcode ...")
+      sample_meta$index_2 <- chartr("ATGC", "TACG", stringi::stri_reverse(sample_meta$index_2))
+    } else {
+      stop('ERROR: Reverse index 2 is set to TRUE, but index_2 does nto exists.')
```
Comment (manilkocak) [inline scripts/src/filter_raw_reads.R:95]: typo : nto --> not

File: scripts/src/filter_raw_reads.R [hunk 7]
```diff
@@ -38,47 +73,77 @@ suppressPackageStartupMessages(library(sets))
...
+  # Create "profile_id" column.
   sample_meta %<>% tidyr::unite('profile_id', all_of(id_cols), sep=':', remove=F, na.rm=F)
+  # Join cell_set_meta and cell_line_meta. The cell_set can be a name "P939" or a list of LUAs.
   template= sample_meta %>% dplyr::left_join(cell_set_meta, by= 'cell_set') %>%
     dplyr::mutate(members= ifelse(is.na(members), str_split(cell_set, ';'), str_split(members, ';'))) %>% 
-    tidyr::unnest(cols= c(members)) %>%
-    dplyr::left_join(cell_line_meta, by= dplyr::join_by('members'=='LUA'), relationship= 'many-to-one')
+    tidyr::unnest(cols= members) %>%
+    dplyr::left_join(cell_line_meta, by= dplyr::join_by('members'=='LUA'), relationship= 'many-to-one') %>%
+    dplyr::distinct() # to remove duplicate cell lines - cell sets sometime have the same cell line repeated.
```
Comment (ashishge1) [inline scripts/src/filter_raw_reads.R:144]: if there are duplicate cell lines in the same cell_set (not project), that could be an issue. (especially if the duplicates were from different pools).
is it possible to print out a warning and the duplicate entries here before doing the distinct operation?

**Review by ashishge1 — APPROVED** [review]
Everything look good. 
Just left the one comment on lines 142 in filter_raw_reads.R where duplicate cell lines in cell set are removed from the template.
It would be nice if a warning message could be printed out with the information of the duplicate cell lines before the distinct is applied!

Thanks so much for these changes Yuhjong!

File: scripts/src/filter_raw_reads.R (hunk 7 above)
Comment (yuhjong24) [reply scripts/src/filter_raw_reads.R]: Added a validator for cell_set_meta

**Pull request closed without merging**